src/examples/libpmem2/unsafe_shutdowns.c

```diff
@@ -0,0 +1,429 @@
+		fprintf(stderr, "%s\n", pmem2_errormsg());
+		exit(1);
+	}
+
+	/*
+	 * 5. Flush
+	 * FAIL: BAD UUID, BAD USC (may be good if last file), BAD checksum
+	 */
+	flush(mapping_addr, mapping_size);
+
+	/*
+	 * 6. Calculate checksum
+	 * FAIL: BAD UUID, BAD USC (may be good if last file), BAD checksum
+	 */
+	checksum_compute(sds, SDS_LEN_NO_CS, &sds->checksum);
+
+	/*
+	 * 7. Flush
+	 * FAIL: BAD UUID, BAD USC (may be good if last file), GOOD checksum
+	 */
+	flush(mapping_addr, mapping_size);
+
+	/*
+	 * 8. Set flag
+	 * FAIL: GOOD UUID, GOOD USC, GOOD || BAD checksum,
+	 * FLAG == CLEAR_FLAG || FLAG == DIRTY_FLAG
+	 */
+	sds->flag = DIRTY_FLAG;
+
+	/*
+	 * 9. Flush
+	 * FAIL: GOOD UUID, GOOD USC, BAD checksum, FLAG == 1
+	 */
+	flush(mapping_addr, mapping_size);
+
+	/*
+	 * 10. Calculate checksum
+	 * FAIL: GOOD UUID, GOOD USC, BAD checksum, FLAG == DIRTY_FLAG
+	 */
+	checksum_compute(sds, SDS_LEN_NO_CS, &sds->checksum);
+
+	/*
+	 * 11. Flush
+	 * FAIL: GOOD UUID, GOOD USC, GOOD checksum, FLAG == DIRTY_FLAG
+	 */
+	flush(mapping_addr, mapping_size);
+
+	/*
+	 * 12. Save something else to the pool and flush
+	 * FAIL: GOOD UUID, GOOD USC, GOOD checksum, FLAG == 1
+	 */
+	char *string_to_save = "hello, persistent memory";
+	strcpy((char *)mapping_addr + SDS_LEN, string_to_save);
+	flush(mapping_addr, mapping_size);
+
+	/*
+	 * 13. Clear flag
+	 * FAIL: GOOD UUID, GOOD USC, GOOD || BAD checksum,
+	 * FLAG == 1 || FLAG == 0
+	 */
+	sds->flag = CLEAR_FLAG;
+
+	/*
+	 * 14. Flush
+	 * FAIL: GOOD UUID, GOOD USC, BAD checksum, FLAG == 0
+	 */
+	flush(mapping_addr, mapping_size);
+
+	/*
+	 * 15. Calculate checksum
+	 * FAIL: GOOD UUID, GOOD USC, BAD checksum, FLAG == CLEAR_FLAG
+	 */
+	checksum_compute(sds, SDS_LEN_NO_CS, &sds->checksum);
+
+	/*
+	 * 16 Flush
+	 * FAIL: GOOD UUID, GOOD USC, GOOD checksum, FLAG == 0
+	 */
+	flush(mapping_addr, mapping_size);
+
+	/* 17. File unmapping */
+	cleanup(fd, cfg, map);
+}
+
+/*
+ * read_data -- reading mapping and checking/fixing integrity of the data
+ */
+static void
+read_data(int *fd, char *file, struct pmem2_config *cfg,
+		struct pmem2_map *map)
+{
+	void *mapping_addr = NULL;
+	size_t mapping_size;
+	pmem2_flush_fn flush;
+
+	prepare_environment(fd, file, cfg, &map, &mapping_addr, &mapping_size);
+	struct shutdown_state *sds = (struct shutdown_state *)mapping_addr;
+	flush = pmem2_get_flush_fn(map);
+
+	/* stack variables used to compare data with the *sds */
+	uint64_t usc; /* unsafe shutdown count */
+	uint8_t uuid[UUID_LEN]; /* UID */
+	uint64_t checksum;
+
+	/* let's get dimm's UUID, USC and get *sds checksum */
+	size_t len = SDS_LEN_NO_CS;
+	if (pmem2_get_device_id(cfg, (char *)uuid, &len)) {
+		fprintf(stderr, "%s\n", pmem2_errormsg());
+		exit(1);
+	}
+	if (pmem2_get_device_usc(cfg, &usc)) {
+		fprintf(stderr, "%s\n", pmem2_errormsg());
+		exit(1);
+	}
+	checksum_compute(sds, SDS_LEN_NO_CS, &checksum);
+
+	/* UUID && USC == GOOD */
+	if (sds->uuid == uuid && sds->usc == usc) {
```

Comment:
this can be radically simplified using a couple of functions with good names, like:
```
if (!pool_is_corrupted()) {
    if (pool_is_clear())
        return 0;
    else
        return pool_reinit();
} else if (pool_is_moved()) {
    return pool_fix_location();
} else {
    return pool_reinit();
}
```
At least I think so.